aips/move-spec-verifier.md

```diff
@@ -0,0 +1,114 @@
+
+## Impact
+
+As the proposed toolset verifies Move specifications, the target audience will be mainly the Move developers.
+
+Smart contract developers will be able to perform additional action - contract specification verification, allowing them to improve the quality of their contracts before deploying them on the network.
+
+## Alternative solutions
+
+There is no standard solution for testing Move specifications. No other tool can also generate mutants from the source code for the Move programs.
+
+Bytecode mutation can be an alternative solution for generating mutations for the Move smart contracts. However, such an approach can be less comfortable for the developers, as it produces mutations that must be decompiled to the source code to see the differences between them and the original code. Decompiling may introduce some noise (like omitting comments), making it harder to track the differences.
+
+Similar tools for testing specifications are available for other languages, such as Solidity. There are at least two notable tools: Certora Gambit and SuMo-SOlidity-MUtator.
+
+## Specification
+
+The specification verification tool is placed inside the `aptos-core` repository, providing two additional `aptos` subcommands - `spec-verify` and `mutate`.
+
+The `mutate` command is responsible for generating mutants from the Move source code. The Move mutator tool takes the source code as input and generates mutants by introducing small changes in the code. The changes are based on the mutation operators. Basically, the `mutate` command traverses the source files and applies mutation operators (making new mutants) to all possible places. Each mutant is saved in a separate file - one mutation per file. Before saving, the mutant can be checked by trying to compile it to avoid passing to the Move Prover invalid mutants.
+
+1. Binary operator replacement - replaces binary operators with other binary operators. For example, the `+` operator can be replaced with the `-` operator. Operators are grouped into the following categories (operators are replaced within the same category. For example, the `+` operator can be replaced with the `-` operator but not the `<<` operator):
+   - arithmetic operators: `+`, `-`, `*`, `/`, `%`
+   - bitwise operators: `&`, `|`, `^`
+   - shifts: `<<`, `>>`
+   - comparison operators: `==`, `!=`, `<`, `>`, `<=`, `>=`
+   - logical operators: `&&`, `||`
+2. Unary operator replacement - this mutation operator replaces unary operators with other unary operators. For example, the `!` operator can be replaced with the space.
+3. Type replacement - replaces types with other types. For example, the `u8` type can be replaced with the `u64` type.
```

Comment:
This could fail type check ... same with 4, 6.

Reply:
Indeed, type replacement may fail type check. I'm not sure how many cases would fail and how many could survive - I find it interesting to check it. We can always remove that operator if it does not work well.
4th point is a literal replacement - I believe it shouldn't fail when we replace them inside one type of group.
6th point of the list is a return value replacement. Still, if we could replace the returned value with the literal of the same type, it should compile properly.
We can always discuss and replace those selected operators with others to be built in the M3.